presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CombineApproxPercentileFunctions.java

```diff
@@ -0,0 +1,278 @@
+
+/**
+ * For multiple approx_percentile() function calls on the same column with different percentile arguments, combine them to one call on an array of percentile arguments.
+ * <p>
+ * From:
+ * <pre>
+ * - Aggregation (approx_percentile(col, 0.2), approx_percentile(col, 0.8)
+ * </pre>
+ * To:
+ * <pre>
+ * - Project (approx_percentile_results[1], approx_percentile_results[2])
+ *   - Aggregation (approx_percentile_results <- approx_percentile(col, array)
+ *     - Project (col <- col, array <- [0.2, 0.8])
+ * </pre>
+ * <p>
+ */
+
+public class CombineApproxPercentileFunctions
+        implements Rule<AggregationNode>
+{
+    private static final String APPROX_PERCENTILE = "approx_percentile";
+    // Limit specified in `ArrayConstructor` function.
+    private static final int ARRAY_SIZE_LIMIT = 254;
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public CombineApproxPercentileFunctions(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+    }
+
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(CombineApproxPercentileFunctions::hasMultipleApproxPercentile);
+
+    private static boolean hasMultipleApproxPercentile(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations().values().stream()
+                .filter(agg -> agg.getCall().getDisplayName().equals(APPROX_PERCENTILE)).count() > 1;
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isCombineApproxPercentileEnabled(session);
+    }
+
+    // Get the position of the percentile argument in arguments
+    private static int getPercentilePosition(FunctionHandle functionHandle)
+    {
+        // approx_percentile(x, percentage) -> arguments.size() == 2
+        // approx_percentile(x, percentage, accuracy) -> arguments.size() == 3 && arguments.get(1) is Double
+        // approx_percentile(x, w, percentage) -> arguments.size() == 3 && arguments.get(1) is BigInt
+        // approx_percentile(x, w, percentage, accuracy) -> arguments.size() == 4
+        if (functionHandle instanceof BuiltInFunctionHandle) {
+            List<TypeSignature> argumentTypes = ((BuiltInFunctionHandle) functionHandle).getSignature().getArgumentTypes();
+            if (argumentTypes.size() == 2
+                    || (argumentTypes.size() == 3 && argumentTypes.get(1).getBase().equals(StandardTypes.DOUBLE))) {
+                return 1;
+            }
+            else if (argumentTypes.size() == 4
+                    || (argumentTypes.size() == 3 && argumentTypes.get(1).getBase().equals(StandardTypes.BIGINT))) {
+                return 2;
+            }
+        }
+
+        checkState(false);
+        return -1;
```

Comment:
Do we need return -1 after checkstate?

Reply:
Put it here as there will be lint error otherwise.

Reply:
OK since we are not removing this function. Let's clean this up :)

Firstly `functionHandle instanceof BuiltInFunctionHandle` - make it a checkState and remove the if.

Then also make
`else if (argumentTypes.size() == 4
                    || (argumentTypes.size() == 3 && argumentTypes.get(1).getBase().equals(StandardTypes.BIGINT))) {`

into a checkState.

Then there is no need for the final checkstate or return -1